src/main/kotlin/io/kotest/assertions/arrow/core/Either.kt

```diff
@@ -0,0 +1,71 @@
+package io.kotest.assertions.arrow.core
+
+import arrow.core.Either
+import arrow.core.Either.Left
+import arrow.core.Either.Right
+import io.kotest.property.Arb
+import io.kotest.property.arbitrary.choice
+import io.kotest.property.arbitrary.map
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.contract
+
+/**
+ * smart casts to [Either.Right] and fails with [failureMessage] otherwise.
+ * ```kotlin
+ * import arrow.core.Either.Right
+ * import arrow.core.Either
+ * import io.kotest.assertions.arrow.core.shouldBeRight
+ *
+ * fun main() {
+ *   //sampleStart
+ *   fun squared(i: Int): Int = i * i
+ *   val result = squared(5)
+ *   val either = Either.conditionally(result == 25, { result }, { 25 })
+ *   val a = either.shouldBeRight { "5 * 5 == 25 but was $it " }
+ *   val smartCasted: Right<Int> = either
+ *   //sampleEnd
+ *   println(smartCasted)
+ * }
+ * ```
+ */
+@OptIn(ExperimentalContracts::class)
+fun <A, B> Either<A, B>.shouldBeRight(failureMessage: (A) -> String = { "Expected Either.Right, but found Either.Left with value $it" }): B {
+  contract {
+    returns() implies (this@shouldBeRight is Either.Right<B>)
```

Comment:
Did they just add smart casting with type parameters ?

Reply:
I think this was always available.

Reply:
No, you never used to be able to do with type parameters.